### PR TITLE
Normalize transactions

### LINKED ISSUE #3843: Can't call smart contract method through Metamask 4.5
I follow this example:
https://medium.com/metamask/calling-a-smart-contract-with-a-button-d278b1e76705

It works in Metamask 4.4

But Metamask upgrade to 4.5 yesterday, and things broken.
when calling smart contract method, Metamask will say insufficient balance,
but I do have the eths.

I wrote an working example online: https://www.obeobeko.com/dapp_test.html

```js
var Web3 = require('web3');
var Eth = require('ethjs-query');
var EthContract = require('ethjs-contract');

var accounts;
var account;
var eth;
var eth_contract;

async function waitForTxToBeMined (txHash) {
  let txReceipt
  while (!txReceipt) {
    try {
      txReceipt = await eth.getTransactionReceipt(txHash);
      console.log(txReceipt);
    } catch (err) {
      return indicateFailure(err)
    }
  }
}

function listenForClicks(simpleStorage) {
    var button = document.querySelector('button.set');
    if (button != undefined) {
        button.addEventListener('click', function() {
            var value = document.getElementById('data-value').value;
            simpleStorage.set(value, {from: account})
            .then(function (txHash) {
              console.log('Transaction sent')
              console.dir(txHash)
              waitForTxToBeMined(txHash)
            })
            .catch(function(error) {
                // rejection
                console.log('reject');
                console.debug(error);
            });
        });
    }
}

window.StartApp = {
  start: async function() {

    eth = new Eth(web3.currentProvider);
    eth_contract = new EthContract(eth);

    var self = this;

    // Get the initial account balance so it can be displayed.
    web3.eth.getAccounts(function(err, accs) {
      if (err != null) {
        console.log("There was an error fetching your accounts.");
        return;
      }

      if (accs.length == 0) {
        console.log("Couldn't get any accounts! Make sure your Ethereum client is configured correctly.");
        return;
      }

      accounts = accs;
      account = accounts[0];
      console.log(account);

      web3.eth.getBalance(account, function(err, balance) {
        console.log(balance);
      });
    });

    const abi = [ { "constant": true, "inputs": [], "name": "data", "outputs": [ { "name": "", "type": "uint256" } ], "payable": false, "stateMutability": "view", "type": "function" }, { "anonymous": false, "inputs": [ { "indexed": true, "name": "_from", "type": "address" }, { "indexed": false, "name": "value", "type": "uint256" } ], "name": "Set", "type": "event" }, { "constant": false, "inputs": [ { "name": "x", "type": "uint256" } ], "name": "set", "outputs": [], "payable": false, "stateMutability": "nonpayable", "type": "function" } ];
    const addresss = '0x06e1c13546e04514a0cf8d842216a84745ac317a';
    const SimpleStorage = eth_contract(abi);
    const simpleStorage = SimpleStorage.at(addresss);
    // Listen to clicks from a <button> that trigger a function call of contract

    listenForClicks(simpleStorage);
  },
  getBalance: async function(address) {
    return await web3.eth.getBalance(address);
  }
};

window.addEventListener('load', function() {
  // Checking if Web3 has been injected by the browser (Mist/MetaMask)
  if (typeof web3 !== 'undefined') {
    // Use Mist/MetaMask's provider
    window.web3 = new Web3(web3.currentProvider);
    StartApp.start();
  }
});
```

```html
<input id="data-value" type="text" placeholder="Enter a number">
<button class="set">Set Data!</button>
<script src="/js/dapp_test.js"></script>
```
Anyone to solve this issue? Thanks!

### PR DESCRIPTION
**Pull request by frankiebee**
Fixes #3843

### CODE REVIEW

**Comment by frankiebee** [comment]
fixes #3843

File: app/scripts/controllers/transactions.js [hunk 1]
```diff
@@ -314,6 +314,56 @@ module.exports = class TransactionController extends EventEmitter {
 //           PRIVATE METHODS
 //
 
+  _normalizeTxParams (txParams) {
+    delete txParams.chainId
+
+    if ( !txParams.to ) delete txParams.to
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: please use brackets when an `if` has and `else`

File: app/scripts/controllers/transactions.js [hunk 2]
```diff
@@ -185,7 +185,8 @@ module.exports = class TransactionController extends EventEmitter {
 
   async addUnapprovedTransaction (txParams) {
     // validate
-    await this.txGasUtil.validateTxParams(txParams)
+    await this._validateTxParams(txParams)
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: remove the await if we make `_validateTxParams` sync

File: app/scripts/migrations/024.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  const newState = state
+  const transactions = newState.TransactionController.transactions
+
+  newState.TransactionController.transactions = transactions.map((txMeta, _, txList) => {
+    if (
+      txMeta.status === 'unapproved' &&
+      txMeta.txParams &&
+      txMeta.txParams.from
+      ) {
+      txMeta.txParams.from = txMeta.txParams.from.toLowerCase()
```
Comment (kumavis) [inline app/scripts/migrations/024.js]: 👍

File: app/scripts/controllers/transactions.js [hunk 4]
```diff
@@ -314,6 +314,56 @@ module.exports = class TransactionController extends EventEmitter {
...
+    if (!txParams.data) delete txParams.data
+    else txParams.data = ethUtil.addHexPrefix(txParams.data)
+
+    if (txParams.value) txParams.value = ethUtil.addHexPrefix(txParams.value)
+
+    if (txParams.gas) txParams.gas = ethUtil.addHexPrefix(txParams.gas)
+    if (txParams.gasPrice) txParams.gas = ethUtil.addHexPrefix(txParams.gas)
+  }
+
+  async _validateTxParams (txParams) {
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: maybe doesnt need to be async?

**Review by kumavis — CHANGES_REQUESTED** [review]

**Comment by kumavis** [comment]
Awesome thanks @frankiebee !

**Review by kumavis — APPROVED** [review]

**Pull request merged**